### DISCUSSION TITLE
.nef image to jpg conversion issue.

### DISCUSSION THREAD

**Discussion by sruthi09I** [ImageMagick/ImageMagick#3133] [General]
Hi All,

I have a raw image .nef file which I am trying to convert to jpg file. Converted jpg file color is not as expected, it is darker than the original image. 

On adding an extra parameter(-e) to delegates.xml file in ImageMagick-7.0.3 version as below, the issue is resolved:
<delegate decode="dng:decode" stealth="True" command="dcraw.exe -e -6 -W -O "%u.ppm" "%i""/>

But tried adding the same parameter -e in later imagemagick versions, the issue still exits. Tried in imagemagick versions - ImageMagick-7.0.6-Q16, 7.0.8-3 Q16 and ImageMagick 7.0.10-10 Q16.

Do we require to add any other parameters in latest imagemagick versions? If yes can you please the details.

Command used to convert raw image to jpg format:
magick.exe nef:C:\ASSETS\cust\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\ASSETS\cust\OriginalAsset-S.JPG

Verbose output with IM 7.0.3 version(working):
dcraw.exe -6 -W -O "C:/Users/sinampud/AppData/Local/Temp/magick-30052jo1esoFwvcjL.ppm" "C:/Users/sinampud/AppData/Local/Temp/magick-30052R4Ji-g3TlcFG"
C:/Users/sinampud/AppData/Local/Temp/magick-30052jo1esoFwvcjL.ppm PPM 2868x4312 2868x4312+0+0 16-bit sRGB 74.2MB 0.063u 0:00.074
nef:C:\ASSETS\cust\OriginalAsset.nef=>C:/Users/sinampud/AppData/Local/Temp/magick-30052jo1esoFwvcjL.ppm NEF 2868x4312 2868x4312+0+0 16-bit sRGB 74.2MB 0.016u 0:00.012
nef:C:\ASSETS\cust\OriginalAsset.nef=>C:\ASSETS\cust\OriginalAsset-S.JPG NEF 2868x4312=>511x768 511x768+0+0 16-bit sRGB 152KB 0.031u 0:00.015

Verbose output with later IM versions(not working):
nef:C:\ASSETS\cust\OriginalAsset.nef=>C:\ASSETS\cust\OriginalAsset.nef NEF 2868x4312 4320x2868+0+0 16-bit sRGB 22.7287MiB 3.359u 0:01.213
nef:C:\ASSETS\cust\OriginalAsset.nef=>C:\ASSETS\cust\OriginalAsset-S.JPG NEF 2868x4312=>511x768 770x511+0+0 16-bit sRGB 155744B 0.031u 0:00.017

Attached the file which contains details about the .net image.
[metadata.txt](https://github.com/ImageMagick/ImageMagick/files/5830041/metadata.txt)

**Comment by snibgo** [comment]
The "-e" option to dcraw tells dcraw to extract the embedded JPG image, instead of reading and processing the raw data (white balance, demosaic, etc).

The embedded JPG has been created by the camera as a "prettified" version of the raw data: adjusted tones, increased saturation, sharpened etc. If you read the raw data, you need to do any prettifying yourself.

Your recent version of IM probably uses libraw instead of dcraw. Does "magick -version" have "delegates ... raw"? If so, then it uses libraw. See https://imagemagick.org/script/defines.php and search for "dng:" for relevant defines such as "-define dng:use-auto-wb=true". Specify these before the input NEF file on the command line.

**Comment by sruthi09I** [comment]
Hi,
 Sorry for the delayed response.
 I have tried running the magick -version command. I see delegates being shown, it means libraw is being used.
So i have tried running the command like below:
magick.exe -define dng:use-auto-wb=true nef:C:\ASSETS\cust\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient  C:\ASSETS\OriginalAsset-S.JPG.

But still i did not see the output as expected,  it is darker than the original image.

Can you please let me know if any other options needs to be used.

Thanks in Advance!

**Comment by snibgo** [comment]
> ...  it is darker than the original image.

What "original image"? What are you comparing the IM result to?

If you are comparing with the out-of-camera JPEG then, yes, there will be many differences. The OOC jpeg is made from the same data, but "prettified" with adjusted tones, increased saturation, sharpening and so on.

**Reply by sruthi09I** [reply]
I am comparing the original image with transformed image generated after running the below command:
magick.exe -define dng:use-auto-wb=true nef:C:\ASSETS\cust\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\ASSETS\OriginalAsset-S.JPG
There is a color difference in both the images.
In  ImageMagick-7.0.3 version we don't see any difference in the transformed image. It is exactly same as original image.
On adding an extra parameter(-e) to delegates.xml file in ImageMagick-7.0.3 version as below, it's working fine:
<delegate decode="dng:decode" stealth="True" command="dcraw.exe -e -6 -W -O "%u.ppm" "%i""/>

**Comment by snibgo** [comment]
You haven't answered my question. You have simply repeated that you are comparing the result from IM with some "original image". But what is that "original image"? Where did it come from?

NEF files don't have an "original image". NEF files contain raw values from the camera sensor. They usually also contain a JPEG that the camera has made from the raw values. Perhaps this is what you call the "original image", but this is a processed image, a prettified version.

dcraw's "-e" option extracts that JPEG.

**Reply by sruthi09I** [reply]
Original image was captured from Nikon camera (D850) with .nef extension.  Image with .nef extension is the one i am referring to as original image.
So i'm trying to transform/extract the image to JPEG format(I think is the embedded image that you are referring to) by running the below image magick command:
magick.exe -define dng:use-auto-wb=true nef:C:\ASSETS\cust\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\ASSETS\OriginalAsset-S.JPG..

My expectation is both the images .nef and .jpeg images should have the same color. But with latest imagemagick the .jpeg image is darker than .nef image.

With imagemagick version 7.0.3 both the images .nef and .jpeg images have same color as expected.

The sample that i  have is a customer image, so i will not be able to share the image.

**Reply by sruthi09I** [reply]
I tried for other .nef images. Observed the same issue. 
Please check the attached image files for reference:
Original .nef file : 2311_D056_00003.nef ( Attached this as .jpg extension as the forum doesnot support attaching .nef files. Please convert the .jpg extension to .nef after downloading the file.) 
Jpg generated with Image magick 7.0.3-8 : Converted_with_IM_7.0.3-8.JPG
Jpg generated with Image magick 7.0.10-10 : Converted_with_IM_ 7.0.10-10.JPG

![Converted_with_IM_ 7 0 10-10](https://user-images.githubusercontent.com/77619627/123087117-ce0f3d00-d441-11eb-9ad6-92359e65e36d.JPG)
![Converted_with_IM_7 0 3-8](https://user-images.githubusercontent.com/77619627/123087126-d10a2d80-d441-11eb-8f3b-7b17a8d1d67d.JPG)
![2311_D056_00003](https://user-images.githubusercontent.com/77619627/123087189-e2ebd080-d441-11eb-9c48-a0c87b11b846.jpg)

**Comment by sruthi09I** [comment]
Can you please suggest me if some other options to be added in latest version of image magick to fix the issue with darker image?

**Comment by snibgo** [comment]
Your command was:
```
magick.exe -define dng:use-auto-wb=true nef:C:\ASSETS\cust\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\ASSETS\OriginalAsset-S.JPG
```
"-intent relative" is only relevant when you use the "-profile" operation to change colorspace. As you have no "-profile", "-intent relative" does nothing.

"-sample" is faster than "-resize", but the result is lower quality.

"-strip" removes any embedded ICC profiles. This is usually a bad idea. However, when IM converts from a raw file, it doesn't embed a profile, so this isn't a problem. (I don't know why it doesn't embed a profile. I think it should.)

The small image you show is the OOC JPEG embedded in the NEF file. Nikon does not publish their algorithm that creates the JPEG. I haven't seen a reverse-engineered algorithm.

The major difference between the versions you show isn't in the lightness, but the colour balance. The large image is more yellow (or less blue, which is the same thing). You can adjust colour balance with IM:
```
magick -define dng:use-auto-wb=true 2311_D056_00003.NEF -channel B -evaluate Multiply 0.7 +channel c.tiff
```
This gives a fairly pleasing appearance to the people.

**Reply by sruthi09I** [reply]
Thank you for providing the details.
I tried the below IM command as mentioned. Stiil i see the color difference in the image, same as earlier. 
magick -define dng:use-auto-wb=true C:\ASSETS\nef\2311_D056_00003.nef -channel B -evaluate Multiply 0.7 +channel C:\ASSETS\nef\2311_D056_00003.jpeg
Please refer to attached image. Do i need to include any other [[options?]]
![2311_D056_00003](https://user-images.githubusercontent.com/77619627/123634062-99332980-d837-11eb-8420-ab17c0acd812.jpeg)

**Reply by coffeesmell** [reply]
try using camera white balance...not the auto-white-balance...

add this parameter in your command `-define dng:use-camera-wb=true`
this works for me in latest IM versions

**Comment by snibgo** [comment]
You can multiply the RGB channels by anything you want. For example, you can change the image to have the same overall lightness and colour balance of the OOC jpeg ("ooc.jpg"). Windows BAT script, and assuming that "magick" is HDRI:
```
magick ^
  -define dng:use-auto-wb=true ^
  2311_D056_00003.nef ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ooc.jpg ^
  ( -clone 0,1 ^
    -scale "1x1^!" ^
    -compose DivideDst -composite ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -delete 1 ^
  -compose Multiply -composite ^
  out.jpg
```
The OOC JPEG isn't an "original" image. It is merely one possibility.

**Reply by sruthi09I** [reply]
Tried running the bat script as mentioned above, iam getting the below error: magick: unable to open image 'ooc.jpg': No such file or directory @ error/blob.c/OpenBlob/3496.
Can you please provide more details on what does ooc.jog means? and how to pass it in the command?

**Comment by snibgo** [comment]
Above, you showed the out-of-camera JPEG. It is named "123087126-d10a2d80-d441-11eb-8f3b-7b17a8d1d67d.JPG". For simplicity, I renamed it "ooc.jpg". It doesn't matter what it is named. Just use its name instead of "ooc.jpg".

**Reply by sruthi09I** [reply]
Hi Snibgo, 
As suggested in the forum, i have tried running the below imagemagick command in latest version image magick(ImageMagick-7.1.0-Q16-HDRI) and i was able to see the better output.
magick.exe -define dng:use-camera-wb=true nef:C:\Assets\Images\nef\OriginalAsset.nef -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\Assets\OriginalAsset-S.JPG.

I was having a similar issue for .DNG file Not sure if i need to open a new ticket for this.
The dng image is having a blue color transparent layer in the original image.Bout on trying to convert to jpg image with latest version of image magick (ImageMagick-7.1.0-Q16-HDRI), te blue layer is skipped of and the actual background image is shown.
I have tried the below image magick commands:
magick.exe -define dng:use_camera_wb=true nef:C:\Assets\Images\dng\blue1.dng -intent relative -sample "1024x768>" -quality 95 -colorspace sRGB -strip -auto-orient C:\Assets\Images\dng\Output\DNG-output-color-cm-wb.JPG
Also tried all the different possible optionsas below, but didn't work. Can you please suggest if i need to use some other option.
-define dng:use-camera-wb=true
-define dng:use-auto-wb=true
-define dng:output-color= 0 to 6
-define dng:use-auto-bright=false

Can you please suggest me if there is some other option or way to preserve the blue transparent layer in the transformed jpg image?

**Comment by snibgo** [comment]
> The dng image is having a blue color transparent layer in the original image.Bout on trying to convert to jpg image with latest version of image magick (ImageMagick-7.1.0-Q16-HDRI), te blue layer is skipped of and the actual background image is shown.

Please link to your "blue1.dng". Perhaps the delegate IM uses to read your DNG cannot see the "blue color transparent layer".

DNG files are normally used for images made by digital cameras. Where did the "blue color transparent layer" come from?

JPEG cannot record transparency.

**Reply by sruthi09I** [reply]
Hi Snibgo,
 The image bue1.dng is a customer image, which i may not be able to share in this public forum.
Can you please suggest me some alternate way to share the image, so that it will be internal and only you can see the image.
Thank You.

**Comment by snibgo** [comment]
A DNG file usually contains a "digital negative", a raw (mosaic) file from a digital camera. It may also contain JPEG versions of the same image, possibly downscaled.

But a DNG is really a TIFF file that contains particular data. As such, it can contain any number of images that may not be related.

Most software will try to read the main image, and demosaic it. So I suggest you copy blue1.dng to a tiff file, eg blue1_test.tiff. Then try to open both files with different software.

What does "magick blue1.dng info:" say about the image(s)?

What does "magick blue1_test.tiff info:" say about the image(s)?

Try opening both files in Gimp. What does it say about the images?

What does exiftool say about the files?

What does dcraw say about blue1.dng? How about other raw readers such as darktable and RawTherapee and libraw tools?

EDIT to add: You could ask your customer what software created the blue transparent layer. It may be in private metadata that can only be read by that software.

**Reply by sruthi09I** [reply]
Please check the attached files containing the data requested abobe after running imagemagick and exitool on both original dng and tiff converted from dng file.
I was able to see the blue transparent layer with windows photo viewer as well. @
[magick_info.txt](https://github.com/ImageMagick/ImageMagick/files/6999933/magick_info.txt)
[Exiftool Output.txt](https://github.com/ImageMagick/ImageMagick/files/6999934/Exiftool.Output.txt)
Let me know if i need to share any other information. I can also share the image if required but not in forum.

**Comment by snibgo** [comment]
I can't see anything in those text files about another layer.

> I was able to see the blue transparent layer with windows photo viewer as well. 

Does that software process the raw image (demosaic etc) or does it simply show the embedded JPG?

**Reply by sruthi09I** [reply]
Hi Snibgo,
 I have no idea about this "Does that software process the raw image (demosaic etc) or does it simply show the embedded JPG?" . Don't know how to check this.
   Can you please share your mail id or some temporary mail id to share the image? I feel it will easy for you to understand by looking at the image.

**Comment by snibgo** [comment]
> I have no idea about this "Does that software process the raw image (demosaic etc) or does it simply show the embedded JPG?" . Don't know how to check this.

Your exiftool output shows that the raw file contains "Jpg From Raw" and "Preview Image" and "Thumbnail TIFF". You can extract these with exiftool. For help, type "exiftool". Do these have the blueness that you are looking for? Perhaps your customer wants one of these. But these will be lower quality than the processed raw image.

Then you can process (demosaic etc) the raw image with dcraw or similar. I suppose this doesn't have the blueness.

If the processed raw image doesn't have the blueness, but a JPEG does, there are techniques for shifting colours to make the first match the second. For example, see [Gain and bias](http://im.snibgo.com/gainbias.htm).

**Reply by sruthi09I** [reply]
I tried running running exiftool with options "Jpg From Raw" and "Preview Image" and "Thumbnail TIFF".
Observed all the 3 output jpegs have the blue transparent layer over it.

**Reply by sruthi09I** [reply]
Hi Snibgo,
 Can you please let me know the command to be used in imagemagick to retrieve the image with blue transparent layer over it?

**Comment by snibgo** [comment]
As I have said, it appears that IM can't see a blue transparent layer.

**Reply by sruthi09I** [reply]
Ok.Thank you so much Snibgo for providing the details.

I'm facing an issue with the output generated for nef file which we have discussed earlier? 
If i run the command  as below with -define tag before the input file path i am able to see the desired output(Image with correct color):
magick **-define dng:use-camera-wb=true** nef:C:\Assets\Images\nef\Changedname.nef  -intent relative -sample 1024x768 -quality 95  -colorspace sRGB -strip -auto-orient -density 72 C:\Assets\Images\nef\Changedname-S1.JPG

But if i run the command as below, with -define tag followed by input file path i'am not getting the desired output. I see the output image is darker.
magick nef:C:\Assets\Images\nef\Changedname.nef **-define dng:use-camera-wb=true**  -intent relative -sample 1024x768 -quality 95 -colorspace sRGB -strip -auto-orient -density 72 C:\Assets\Images\nef\Changedname-S.JPG

Is the order of -define tag important? Can you please help me with some clarification on the order considered for nef file that we discussed earlier?

**Comment by snibgo** [comment]
Yes, the order is important. IM executes the options in the order you give them. "-define dng:..." affect how the raw file is read, so need to be _before_ the NEF file is read. If you set set afterwards, they have no effect.
